thrift/arg2/kv_iterator.go

```diff
@@ -0,0 +1,84 @@
+// Package arg2 contains tchannel thrift Arg2 interfaces for external use.
+//
+// These interfaces are currently unstable, and aren't covered by the API
+// backwards-compatibility guarantee.
+package arg2
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+// KeyValIterator is a iterator for reading tchannel-thrift Arg2 Scheme,
+// which has key/value pairs (k~2 v~2).
+// NOTE: to be optimized for performance, we try to limit the allocation
+// done in the process of iteration.
+type KeyValIterator struct {
+	arg2Payload           []byte
+	leftPairCount         int
+	keyOffset             int
+	valueOffset, valueLen int
+}
+
+// NewKeyValIterator inits a KeyValIterator with the buffer pointing at
+// start of Arg2. Return io.EOF if no iterator is available.
+// NOTE: tchannel-thrift Arg Scheme starts with number of key/value pair.
+func NewKeyValIterator(arg2Payload []byte) (KeyValIterator, error) {
+	if len(arg2Payload) < 2 {
+		return KeyValIterator{}, io.EOF
+	}
+
+	return KeyValIterator{
+		valueOffset:   2, // nh has 2B offset
+		leftPairCount: int(binary.BigEndian.Uint16(arg2Payload[0:2])),
+		arg2Payload:   arg2Payload,
```

Comment:
optionally: instead of setting `valueOffset` to `2`, you can leave it at default 0, and pass in `arg2Payload[2:]`